sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala

```diff
@@ -323,22 +323,26 @@ case class WrapOption(child: Expression)
 }
 
 /**
- * A place holder for the loop variable used in [[MapObjects]].  This should never be constructed
- * manually, but will instead be passed into the provided lambda function.
+ * A place holder for the loop variable used in [[MapObjects]].
  */
-case class LambdaVariable(value: String, isNull: String, dataType: DataType) extends Expression {
+case class LoopVar(value: String, isNull: String, dataType: DataType) extends LeafExpression
+  with Unevaluable {
 
-  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String =
-    throw new UnsupportedOperationException("Only calling gen() is supported.")
+  override def nullable: Boolean = true
 
-  override def children: Seq[Expression] = Nil
-  override def gen(ctx: CodeGenContext): GeneratedExpressionCode =
+  override def gen(ctx: CodeGenContext): GeneratedExpressionCode = {
     GeneratedExpressionCode(code = "", value = value, isNull = isNull)
+  }
+}
 
-  override def nullable: Boolean = false
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported.")
+object LoopVar {
+  private val curId = new java.util.concurrent.atomic.AtomicInteger()
 
+  def apply(dataType: DataType): LoopVar = {
+    val loopValue = "MapObjects_loopValue" + curId.getAndIncrement()
+    val loopIsNull = "MapObjects_loopIsNull" + curId.getAndIncrement()
+    LoopVar(loopValue, loopIsNull, dataType)
```

Comment:
Here is a little hacky, but I wanna argue that it's way less hacky than the previous implementation and I make the variable name verbose to make it less possible to get conflicted.